[PM-5971] Fix Payment Method Warning Bugs

amorask-bitwarden opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [X] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

<!--Describe what the purpose of this PR is. For example: what bug you're fixing or what new feature you're adding-->

This PR refactors the implementation of the payment method warning banners to continuously poll the `/organizations/{organizationId}/billing-status` endpoint for whether the active user's organizations risk subscription failure. It uses the responses from those endpoints to update an `ActiveUserState` observable  used to determine whether or not to display the payment method warning banners.

## Key Code Changes

<!--Explain the changes you've made to each file or major component. This should help the reviewer understand your changes-->
<!--Also refer to any related changes or PRs in other repositories-->

#### Rendering
- **organization-layout.component.*:** Render newly refactored component if FF is on.
- **user-layout.component.*:** Render newly refactored component if FF is on.
- **providers-layout.componen.*:** Render newly refactored component if FF is on. 
- **setup.component.*:** Render newly refactored component if FF is on.
- **payment-method-warning-banners.component.*:** Consume the new `paymentMethodWarningService.paymentMethodWarnings$` observable and display the resulting banners accordingly.

#### Business Logic
- **organization-api.service.*:** Rename endpoint method to target `/billing-status` per https://github.com/bitwarden/server/pull/3790.
- **payment-method-warning.service.*:** Main bulk of the new logic around fetching each organization's billing status on a timer and updating the `ActiveUserState` observable.
- **app.component.ts:** Clear session storage from payment-method-warning.service when user logs out.
- **adjust-payment.component.ts:** Hide the payment method warning banner when a user adds a payment method.
- **feature-flag.enum.ts:** Add new feature flat to control display.

#### Unrelated Refactor
- **organization-billing.service.*:** Move request models to their own file. Refactor existing methods.
- **secrets-manager-trial-billing-step.component.ts:** Use new `organizationBillingService` method name.
- **secrets-manager-trial-free-stepper.component.ts:** Use new `organizationBillingService` method name. 

## Screenshots

<!--Required for any UI changes. Delete if not applicable-->

https://github.com/bitwarden/clients/assets/144709477/f23e89e4-e1d9-4416-9912-8a9db0b069b6



## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

cyprain-okeke commented on libs/common/src/billing/services/payment-method-warning.service.ts:
```diff
@@ -0,0 +1,97 @@
+    await this.paymentMethodWarningsState.update((state) => {
+      state ??= {};
+      const current = state[organizationId];
+      if (current === null || current === undefined) {
```
Is (current === null || current === undefined)  necessary since the state is initialized to an empty object if it's null or undefined.

amorask-bitwarden replied:
The `state` is initialized to an empty `Record<string, PaymentMethodWarning>` in the case it's `null` or `undefined`. However, `current` is the value retrieved from `state` using `organizationId` as a key, which can be `null` or `undefined` if that key does not exist or if the value for that key is `null` or `undefined`.

cyprain-okeke commented on libs/common/src/billing/services/payment-method-warning.service.ts:
```diff
@@ -0,0 +1,97 @@
+      filter((state) => state !== null),
+    );
+
+    const latestStatuses$: Observable<OrganizationBillingStatusResponse[]> = defer(() =>
```
If there's no specific reason for using defer, it can be simplified by directly calling this.getOrganizationStatuses()

amorask-bitwarden replied:
I could be wrong as I don't have too much RxJS experience, but I wanted to use this to defer the creation of an observable until it's actually needed as part of the `refreshTimer$` observable on line 33: https://github.com/bitwarden/clients/pull/7923/files/7f8d84077af5aa182ad2c2d1703f9e7940eb3404#diff-86fd279f0a3a40b0fb3fce081c0dc8d31af56c5a74f99afd2c4f5309e14388dfR33

I only want this to be invoked as part of the timer observable.

MGibson1 commented on apps/web/src/app/shared/loose-components.module.ts:
```diff
@@ -110,7 +110,7 @@ import { SharedModule } from "./shared.module";
     PasswordCalloutComponent,
     DangerZoneComponent,
-    PaymentMethodBannersComponent,
+    PaymentMethodWarningBannersComponent,
```
Please create a new module as indicated by comment on lines [97/98](https://github.com/bitwarden/clients/pull/7923/files?file-filters%5B%5D=.html&file-filters%5B%5D=.ts&owned-by%5B%5D=MGibson1&show-deleted-files=true&show-viewed-files=true#diff-26b5f4c999d51c30ac91362870f6a24ef14db4640d00496e1863d521a9e7d18fR97-R98)

amorask-bitwarden replied:
Sorry, missed that comment. Updated here: https://github.com/bitwarden/clients/pull/7923/commits/51282589654967bbc103e2e0aaf4571bd6640622

MGibson1 commented on libs/common/src/billing/services/payment-method-warning.service.ts:
```diff
@@ -0,0 +1,97 @@
+      this.getOrganizationStatuses(),
+    );
+
+    this.everyMinuteStartingNow$.pipe(concatMap(() => latestStatuses$)).subscribe((statuses) => {
```
Don't subscribe in your service.

Instead, you should just expose the `PaymentMethodWarnings` directly, subscribe to a timer in each `app.component`, and request and update of the risk from there. 

You can define an update method in your service here. The update method should call update on the `state`, and trigger api requests _only_ for pulls that are older than some amount of time (probably days or weeks).

MGibson1 reviewed (changes requested):
I haven't had a chance to fully review, but a few items I noticed

MGibson1 commented on apps/web/src/app/app.component.ts:
```diff
@@ -241,6 +247,17 @@ export class AppComponent implements OnDestroy, OnInit {
+      )
+      .subscribe((organizations) =>
+        organizations.forEach((organization) =>
+          this.paymentMethodWarningService.update(organization.id),
```
**blocking**: Don't have promises in subscribe callbacks, use a `concatMap` or `switchMap`, depending on which is appropriate.

amorask-bitwarden replied:
Like this?
```js
this.paymentMethodWarningsRefresh$
      .pipe(
        switchMap(() => this.organizationService.memberOrganizations$),
        switchMap(
          async (organizations) =>
            await Promise.all(
              organizations.map((organization) =>
                this.paymentMethodWarningService.update(organization.id),
              ),
            ),
        ),
        takeUntil(this.destroy$),
      )
      .subscribe();this.paymentMethodWarningsRefresh$
      .pipe(
        switchMap(() => this.organizationService.memberOrganizations$),
        switchMap(
          async (organizations) =>
            await Promise.all(
              organizations.map((organization) =>
                this.paymentMethodWarningService.update(organization.id),
              ),
            ),
        ),
        takeUntil(this.destroy$),
      )
      .subscribe();
```

amorask-bitwarden replied:
Like this?

```js
this.paymentMethodWarningsRefresh$
      .pipe(
        switchMap(() => this.organizationService.memberOrganizations$),
        switchMap(
          async (organizations) =>
            await Promise.all(
              organizations.map((organization) =>
                this.paymentMethodWarningService.update(organization.id),
              ),
            ),
        ),
        takeUntil(this.destroy$),
      )
      .subscribe();
```

MGibson1 commented on apps/web/src/app/app.component.ts:
```diff
@@ -241,6 +247,17 @@ export class AppComponent implements OnDestroy, OnInit {
       new SendOptionsPolicy(),
     ]);
 
+    this.paymentMethodWarningsRefresh$
+      .pipe(
+        switchMap(() => this.organizationService.memberOrganizations$),
```
nice way to accomplish this 👍

MGibson1 commented on libs/common/src/billing/abstractions/organization-billing.service.ts:
```diff
@@ -1,45 +1,10 @@
-  purchaseSubscription: (subscription: SubscriptionInformation) => Promise<OrganizationResponse>;
-
-  startFree: (subscription: SubscriptionInformation) => Promise<OrganizationResponse>;
+  purchaseOrganization: (request: PurchaseOrganizationRequest) => Promise<OrganizationResponse>;
+  startFreeOrganization: (request: StartFreeOrganizationRequest) => Promise<OrganizationResponse>;
```
**question**: is this really related to the banner issues?

amorask-bitwarden replied:
No, unrelated refactoring I forgot to remove; removed here: https://github.com/bitwarden/clients/pull/7923/commits/31cbb066f2ccb2239bd19e9cbf4b360d4fb475a5

MGibson1 commented on libs/common/src/billing/models/domain/payment-method-warning.ts:
```diff
@@ -0,0 +1,6 @@
+  organizationName: string;
+  risksSubscriptionFailure: boolean;
+  acknowledged: boolean;
+  savedAt: number;
```
IMO this is better represented as a `Date` with more active deserializer. It makes intent clearer than a parameter name ever could.

MGibson1 commented on libs/common/src/billing/services/payment-method-warnings.service.ts:
```diff
@@ -0,0 +1,76 @@
+  ) {
+    this.paymentMethodWarningsState = this.stateProvider.getActive(PAYMENT_METHOD_WARNINGS_KEY);
+    this.paymentMethodWarnings$ = this.paymentMethodWarningsState.state$.pipe(
+      filter((state) => state !== null),
```
why strict equal? are you expecting `undefined` to make it through for some reason?

MGibson1 replied:
**question**: Do you mean to not emit until something exists in this state? If so, I think this deserves some special documentation, since that's not really the norm. If you do something like await a value here, rather than just observe/subscribe, you'll lock the promise until something exists locally. If you happen to be offline, that could be a very long time, indeed.

amorask-bitwarden replied:
This was a relic of me trying to navigate how to handle the fact that that the `ActiveUserState` would start out `null` until it was set to an initial value. I no longer have to worry about that because I'm handling it as part of the polling `update` process here: https://github.com/bitwarden/clients/blob/9abbdf0c3e9712be937adc9060aca358c8f87b95/libs/common/src/billing/services/payment-method-warnings.service.ts#L50-L54

I'll remove this filter.

MGibson1 commented on libs/common/src/billing/abstractions/payment-method-warnings-service.abstraction.ts:
```diff
@@ -0,0 +1,11 @@
+import { PaymentMethodWarning } from "../models/domain/payment-method-warning";
+
+export abstract class PaymentMethodWarningsServiceAbstraction {
+  paymentMethodWarnings$: Observable<Record<string, PaymentMethodWarning>>;
+  acknowledge: (organizationId: string) => Promise<void>;
+  addedPaymentMethod: (organizationId: string) => Promise<void>;
+  clear: () => Promise<void>;
+  update: (organizationId: string) => Promise<void>;
```
Please add documentation to this interface

MGibson1 commented on libs/common/src/billing/services/payment-method-warnings.service.ts:
```diff
@@ -0,0 +1,76 @@
+    });
+  }
+
+  async addedPaymentMethod(organizationId: string): Promise<void> {
```
nice!

amorask-bitwarden commented:
@MGibson1 Really appreciate the thorough review; learning a lot with this one that's very helpful. I think I resolved the remaining feedback. I also added tests for the `PaymentMethodWarningsService`.

MGibson1 reviewed (approved):
Nice work 🤗

eliykat commented on apps/web/src/app/components/payment-method-warnings/payment-method-warnings.component.html:
This folder is not owned by any team. I suggest moving this into a billing folder so you can own it, and for better organization - `billing/shared` seems like a good place as that already contains components used in both user & organization contexts.

eliykat commented on apps/web/src/app/components/payment-method-warnings/payment-method-warnings.module.ts:
I suggest using standalone components instead of single component modules.

amorask-bitwarden replied:
@eliykat Sorry bit confused on this one. @MGibson1 pointed out that the `loose-components.module` says to create a new component module here: https://github.com/bitwarden/clients/pull/7923#discussion_r1488536058 so I did so. 

Would I be able to not unnecessarily tack stuff onto the `loose-components.module` with a standalone component?

eliykat replied:
You're right - @MGibson1, why is it better to import a module than to import a standalone component? All the module is doing is adding boilerplate. I take the comment in `LooseComponentsModule` to prohibit **declaring** new components in that module, not **importing** them. (I mean, ideally the list of imports will shrink as declared components are moved elsewhere, but that's a slightly different issue.)

eliykat commented on libs/common/src/admin-console/abstractions/organization/organization-api.service.abstraction.ts:
I suggest creating an `organization-billing-api.service` to house billing-specific methods. Same with the controller on the server side. That way, Billing Team can own the api calls/controllers as well without needing AC Team approval for changes.

eliykat reviewed (changes requested):
I'll leave it to @r-tome to do a full review, just a few comments re: code ownership and file locations.

amorask-bitwarden commented:
Hi @eliykat, thank you for the review. I've addressed all of your feedback here: https://github.com/bitwarden/clients/pull/7923/commits/c14dfc83918b4c839bad531eed9f5f046973f5f0 except for the part about the standalone component where you [asked](https://github.com/bitwarden/clients/pull/7923#discussion_r1496642649) @MGibson1 for input.

In the interest of not holding this open too long, would it be possible to leave the component as is and open a ticket to refactor it into a standalone component if need be down the line? Thanks in advance!

eliykat commented on bitwarden_license/bit-web/src/app/admin-console/providers/providers.module.ts:
```diff
@@ -5,7 +5,7 @@ import { FormsModule } from "@angular/forms";
 import { SearchModule } from "@bitwarden/components";
 import { OrganizationPlansComponent } from "@bitwarden/web-vault/app/billing";
-import { PaymentMethodWarningsModule } from "@bitwarden/web-vault/app/components/payment-method-warnings/payment-method-warnings.module";
+import { PaymentMethodWarningsModule } from "@bitwarden/web-vault/app/billing/shared/payment-method-warnings/payment-method-warnings.module";
```
`billing/shared` has a barrel file (`index.ts`), so you should export the payment method warnings module from there, then import like so:

```ts
import { PaymentMethodWarningsModule } from "@bitwarden/web-vault/app/billing/shared";
```

We don't have any linting around this but we are moving towards using barrel files and we should respect the ones we already have in place.

amorask-bitwarden replied:
@eliykat I tried this and am getting a no-restricted-imports linting error:
![image](https://github.com/bitwarden/clients/assets/144709477/534056ad-f5bb-42d4-a557-1e93480f8a0f)

/web/src/app/billing/shared/index.ts:
```
export * from "./billing-shared.module";
export * from "./payment-method.component";
export * from "./payment.component";
export * from "./sm-subscribe.component";
export * from "./tax-info.component";
export * from "./payment-method-warnings/payment-method-warnings.module";
```

/web/src/app/admin-console/organizations/layout/organization-layout.component.ts:
```
import { PaymentMethodWarningsModule } from "@bitwarden/web-vault/app/billing/shared";
```

eliykat replied:
My bad - yes, if you're within that app you use the relative import (not `@bitwarden/web-vault`). I just meant you drop the filename off the end.

eliykat reviewed (commented):
lgtm, I've left one further comment below but as I'm not a code owner it's not blocking.

eliykat reviewed (dismissed):
LooseComponentsModule question can be resolved at a later date as well.

amorask-bitwarden commented:
@MGibson1 @eliykat @cyprain-okeke @r-tome Thank you all very much for your help and reviews here. I think I've resolved the remainder of the feedback and resolved the conflicts related to the wide-scale layout changes. 

If I could one more round of re-reviews from each of you to wrap this up when you have the chance, I'd greatly appreciate it.

MGibson1 commented on apps/web/src/app/billing/shared/payment-method-warnings/payment-method-warnings.component.ts:
```diff
@@ -18,12 +18,14 @@ export class PaymentMethodWarningsComponent {
-            organizationId,
-            organizationName,
-          })),
+        warnings
```
**non blocking nit**: This ternary is hard to read since it spans so many lines. It looks like the same could have been done by null coalescing `Object.entries(warnings)` to `Object.entries(warnings ?? [])`

amorask-bitwarden replied:
@MGibson1 Thank you, updated.

MGibson1 commented on apps/web/src/app/billing/shared/payment-method-warnings/payment-method-warnings.component.html:
```diff
@@ -1,5 +1,9 @@
 <ng-container *ngFor="let warning of warnings$ | async">
-  <bit-banner bannerType="warning" (onClose)="closeWarning(warning.organizationId)">
+  <bit-banner
+    class="-tw-m-6 tw-flex tw-flex-col tw-pb-6"
```
**warning**:this feels like something that's more appropriate in the component itself @willmartian what are your thoughts?

willmartian replied:
Yes, that sounds good to me @amorask-bitwarden !
